File: 백준/Gold/11000. 강의실 배정/강의실 배정.java
import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;
    static int n, answer;
    static Lecture[] lectures;
    static PriorityQueue<Integer> pq;
    
    static class Lecture {
        int start;
        int end;
        
        public Lecture(int start, int end) {
            this.start = start;
            this.end = end;
        }
    }
    
    static void input() throws IOException {
        n = Integer.parseInt(br.readLine());
        lectures = new Lecture[n];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            lectures[i] = new Lecture(Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken()));
        }
    }
    
    static void process() throws IOException {
        Arrays.sort(lectures, (l1, l2) -> l1.start == l2.start ? l1.end - l2.end : l1.start - l2.start);
        pq = new PriorityQueue<>();
        pq.add(lectures[0].end);
        for (int i = 1; i < n; i++) {
            if (pq.peek() <= lectures[i].start) {
                pq.poll();
            }
            pq.offer(lectures[i].end);
        }
        answer = pq.size();
    }
    
    static void output() throws IOException {
        bw.write(answer + "\n");
        bw.flush();
        bw.close();
    }
    
    public static void main(String[] args) throws IOException {
        input();
        process();
        output();
    }
}

File: 백준/Gold/11000. 강의실 배정/강의실 배정.py
import heapq
import sys

input = sys.stdin.readline

if __name__ == "__main__":
    """
    강의 시작, 종료 시간이 주어지고 필요한 최소 강의실 개수 반환
    """

    # 입력
    n = int(input())
    lectures = []
    for _ in range(n):
        start, end = map(int, input().split())
        lectures.append([start, end])

    # 시작 시간, 종료 시간 순으로 정렬
    lectures.sort(key=lambda x: (x[0], x[1]))

    # 종료 시간을 기준으로 최소 힙
    # 1. 현재 강의의 시작 시간이 힙의 최소 값(가장 먼저 끝나는 강의의 종료 시간)
    # 보다 크거나 같으면 힙에서 삭제
    # 2. 현재 강의의 종료 시간을 힙에 추가
    # 3. 힙의 길이가 필요한 강의실의 개수
    heap = []
    for start, end in lectures:
        if heap and heap[0] <= start:
            heapq.heappop(heap)
        heapq.heappush(heap, end)
    print(len(heap))